### PR TITLE
feat: Added maker balance cache to 0x API db

### PR DESCRIPTION
**Pull request by alexkroeger**
# Description

Added an entity to cache RFQT maker balances to the DB for off-chain order validation in the future.

# Testing Instructions

# Checklist

-   [ ] Update [documentation](https://github.com/0xProject/website/blob/development/mdx/api/index.mdx) as needed. **Website Documentation PR:**
-   [ ] Prefix PR title with `[WIP]` if necessary.
-   [ ] Add tests to cover changes as needed.

### CODE REVIEW

File: migrations/1605637330716-CreateMakerBalanceChainCache.ts [hunk 1]
```diff
@@ -0,0 +1,23 @@
+import { MigrationInterface, QueryRunner, Table } from 'typeorm';
+
+const makerBalanceChainCacheTable = new Table({
+    name: 'maker_balance_chain_cache',
+    columns: [
+        { name: 'token_address', type: 'varchar', isPrimary: true },
+        { name: 'maker_address', type: 'varchar', isPrimary: true },
+        // fields are nullable for the job that adds new rows upon
+        // discovery of a new maker address
+        { name: 'balance', type: 'varchar', isNullable: true },
```
Comment (PirosB3) [inline migrations/1605637330716-CreateMakerBalanceChainCache.ts:10]: oh interesting, so we are keeping this as `varchar`?

File: migrations/1605637330716-CreateMakerBalanceChainCache.ts [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+        { name: 'time_of_sample', type: 'timestamptz', isNullable: true },
```
Comment (PirosB3) [inline migrations/1605637330716-CreateMakerBalanceChainCache.ts:11]: @alexkroeger just curious, should we also add a `time_first_seen` column (that is written the first time a row is added to this table by the API?

File: src/entities/MakerBalanceChainCacheEntity.ts [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+// A table of cached erc20 balances for RFQT market makers
+@Entity({ name: 'maker_balance_chain_cache' })
+export class MakerBalanceChainCache {
+    @PrimaryColumn({ name: 'token_address', type: 'varchar' })
+    public tokenAddress?: string;
+
+    @PrimaryColumn({ name: 'maker_address', type: 'varchar' })
+    public makerAddress?: string;
+
+    @Column({ name: 'balance', type: 'numeric', nullable: true, transformer: BigNumberTransformer })
```
Comment (PirosB3) [inline src/entities/MakerBalanceChainCacheEntity.ts:15]: oh I see, so we are saving this as a string, and the TS code will "deserialize" it as a Big Int?

**Review by PirosB3 — APPROVED** [review]
Added 1 suggestion

File: migrations/1605637330716-CreateMakerBalanceChainCache.ts (hunk 2 above)
Comment (alexkroeger) [reply migrations/1605637330716-CreateMakerBalanceChainCache.ts:11]: Yea I like that 👍

File: src/entities/MakerBalanceChainCacheEntity.ts (hunk 3 above)
Comment (alexkroeger) [reply src/entities/MakerBalanceChainCacheEntity.ts:15]: It TS will read it as a BigNumber, and it's stored as a string. That's how it's done for large values for the meta-tx entities, and I think that's a pretty safe way to handle it.

File: migrations/1605637330716-CreateMakerBalanceChainCache.ts (hunk 1 above)
Comment (alexkroeger) [reply migrations/1605637330716-CreateMakerBalanceChainCache.ts:11]: see other comment.

**Comment by alexkroeger** [comment]
deploy staging

**Comment by alexkroeger** [comment]
deploy production

**Comment by alexkroeger** [comment]
deploy production

**Pull request merged**